### PR TITLE
Adding an example of QLearning for simple navigation

### PR DESCRIPTION
**Pull request by emailweixu**

### CODE REVIEW

File: python/social_bot/util/replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        indices = []
+        is_weights = []
+        while len(indices) < num:
+            idx = random.randint(self._history_length,
+                                 len(self._buffer) - self._future_length - 1)
+            weight = 1.
+            if self._do_not_sample_flags[idx]:
+                continue
+            indices.append(idx)
+            is_weights.append(is_weight)
```
Comment (bayesian) [inline python/social_bot/util/replay_buffer.py]: where does variable is_weight come from? should be weight?

File: examples/train_simple_navigation.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    done = False
+    steps = 0
+
+    while not done:
+        obs = PIL.Image.fromarray(obs).resize(options.resized_image_size,
+                                              PIL.Image.ANTIALIAS)
+        obs = np.transpose(obs, [2, 0, 1])
+        action, q = agent.act(obs, reward)
+        control = [
+            0.02 * (action // options.action_discretize_levels),
```
Comment (bayesian) [inline examples/train_simple_navigation.py]: use env.action_space here.. get rid of hard coded 0.02, also why it is different from 0.4 used in  f_action_feature = lambda _, action: (0.4 * (action // 6), 0.4 * (action % 6))

File: examples/train_simple_navigation.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+from social_bot.util.replay_buffer import PrioritizedReplayBuffer
+from collections import deque, namedtuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+
+
+class Options(object):
```
Comment (bayesian) [inline examples/train_simple_navigation.py:25]: Add comments about algorithms implemented here: double DQN with dueling and priority experience replay. 

also seems very close to baseline.deepq algorithm, how is it different?

File: examples/train_simple_navigation.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        elif random.random() < eps:
+            action = random.randint(0, self._num_actions - 1)
+            q = 0
+        else:
+            input = self._make_input(obs, self._history)
+            input = torch.from_numpy(input).float().to(self._options.device)
+            self._acting_net.eval()
+            with torch.no_grad():
+                q_values = self._acting_net.calc_q_values(input)
+            self._acting_net.train()
```
Comment (bayesian) [inline examples/train_simple_navigation.py]: why it needs to train here?

File: python/social_bot/util/segment_tree.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+
+import operator
+
+
+class SegmentTree(object):
```
Comment (bayesian) [inline python/social_bot/util/segment_tree.py:6]: since we already have dependencies on baseline.. can we just use implementation there.

File: python/social_bot/util/replay_buffer.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                               self._future_length + 1)
+            ]
+            features = f_make_sample(*exps)
+            if not batch_features:
+                for _ in range(len(features)):
+                    batch_features.append([])
+            for feature, batch_feature in zip(features, batch_features):
+                batch_feature.append(feature)
+        for i, batch_feature in enumerate(batch_features):
+            batch_features[i] = np.vstack(batch_feature)
```
Comment (bayesian) [inline python/social_bot/util/replay_buffer.py:123]: feels like there are easier ways to do 110 - 116 here with np functions.. tuple to list then vstack

File: examples/train_simple_navigation.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    "Experience", field_names=["state", "action", "reward", "done"])
+
+
+class QAgent(object):
+    def __init__(self, image_shape, num_actions, options):
+        num_image_channels = image_shape[0]
+        num_input_channels = num_image_channels * (options.history_length + 1)
+        if options.f_action_feature is not None:
+            num_input_channels += len(
+                (options.f_action_feature)(0)) * options.history_length
```
Comment (bayesian) [inline examples/train_simple_navigation.py:191]: this does not seem to match signature of lambda function:
f_action_feature = lambda _, action

**Review by bayesian — CHANGES_REQUESTED** [review]

File: python/social_bot/util/replay_buffer.py (hunk 1 above)
Comment (emailweixu) [reply python/social_bot/util/replay_buffer.py]: Fixed

File: examples/train_simple_navigation.py (hunk 2 above)
Comment (emailweixu) [reply examples/train_simple_navigation.py]: This makes the force in the range of 0 - 0.2. Too big force make it very difficult to control. I changed the force to be -{0.1, -0.05, 0, 0.05, 0.1} is the new version

File: examples/train_simple_navigation.py (hunk 3 above)
Comment (emailweixu) [reply examples/train_simple_navigation.py:25]: Added. Not much difference from baseline. I wrote it to make it a self-contained example and get more understanding of the detail

File: examples/train_simple_navigation.py (hunk 4 above)
Comment (emailweixu) [reply examples/train_simple_navigation.py]: This make _acting_net in train mode. I'll move it learn()

File: python/social_bot/util/segment_tree.py (hunk 5 above)
Comment (emailweixu) [reply python/social_bot/util/segment_tree.py:6]: We will not depend on baseline in the future.

File: examples/train_simple_navigation.py (hunk 2 above)
Comment (emailweixu) [reply examples/train_simple_navigation.py]: the 0.4 in f_action_feature is to normalize action feature in the range of -[1, 1]

File: examples/train_simple_navigation.py (hunk 7 above)
Comment (emailweixu) [reply examples/train_simple_navigation.py:191]: This is because f_action_feature is treated as a class method and '_' is for self.

File: examples/train_simple_navigation.py [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+    def _get_prioritized_replay_beta(self):
+        p = min(1., float(self._total_steps) / self._options.max_steps)
+        return (1 - p) * self._options.prioritized_replay_beta0 + p
+
+    def _get_prioritized_replay_gamma(self):
+        p = min(1., float(self._total_steps) / self._options.max_steps)
+        return (1 - p) * self._options.prioritized_replay_gamma0
+
+    def _update_reward_dist(self):
+        i = len(self._replay_buffer) - 2
```
Comment (bayesian) [inline examples/train_simple_navigation.py:284]: use option.history_length instead of hard coded 2.

**Review by bayesian — APPROVED** [review]
